### PR TITLE
RangeIndex as default index

### PR DESCRIPTION
**Pull request by ARF1**
WIP on making RangeIndex (#9977) the default index

### CODE REVIEW

**Comment by ARF1** [comment]
@jreback @shoyer I tried out what happend making RangeIndex the default index: nothing good. - A whopping 60 errors and 13 test failures. - Most of which meant absolutely nothing to me.

To make the problem more manageable, I started this tree with a minimal `RangeIndex` implementation: no optimization at all, pure fall-back to `Int64Index`. In principle it should behave pretty much like the usual default index - except with a different constructor.

Still I have about 16 errors and 2 failures - most of which I have not idea how they happen. I managed to diagnose and fix one and working on another.

That said, I will really need help debugging this if you really want to make `RangeIndex` the default index: I am fairly limited because I have already exceeded the time I was granted to work on `RangeIndex` by far. I will of course still work on it when I can, but I can no longer do it full-time - which would be required to understand all the pandas internals to fix these widely distributed errors.

Maybe you know another pandas developer would would be willing to help out with this - or take ownership?

**Comment by shoyer** [comment]
Haha, that's not too surprising :). If you keep your branch up to date, I can test it and probably identify at least a few fixes.

**Comment by ARF1** [comment]
@shoyer That would be great. I managed to whittle the issues down to only a few with this minimal `RangeIndex` implementation. Now however I am a bit stumped by the error in `test_set_index2 (pandas.tests.test_frame.TestDataFrame)`. See: https://travis-ci.org/pydata/pandas/jobs/60231079

Also, I am not quite sure what to do with the failure in `test_constructor_mrecarray (pandas.tests.test_frame.TestDataFrame)`. The issue seems to be that a dataframe that originally had a `RangeIndex` returns a dataframe with a `Int64Index` after some operation. That should in principle be accepted as ok, I think, but how would one adjust the test?

**Comment by ARF1** [comment]
@jreback @shoyer Ok, it was truely painfuly but I managed to reduce the issues with the minimal `RangeIndex` implementation as default index to a single test failure. This last one though is stubbornly escaping my analysis.

See: https://travis-ci.org/pydata/pandas/jobs/60815604#L1173

I have narrowed it down to this line: https://github.com/ARF1/pandas/blob/rangeindex_as_default_index/pandas/tools/plotting.py#L1007

Before this line, `data.column.name == 'new'`, after this line, `numeric_data.column.name == None`. This only happens if `RangeIndex` is the default index. For the life of me I cannot figure out why the name is lost in translation. - Even with side-by-side step-by-step debugging of both versions.

Any chance you could help with this?

File: pandas/core/index.py [hunk 1]
```diff
@@ -3299,6 +3316,425 @@ def _wrap_joined_index(self, joined, other):
...
+
+    def _shallow_copy(self, values=None, **kwargs):
+        """ create a new Index, don't copy the data, use the same object attributes
+            with passed in attributes taking precedence """
+        if values is None:
+            return RangeIndex(self.start, self.stop, self.step, 
+                              name=self.name, fastpath=True)
+        else:
+            name = kwargs.get('name', self.name)
+            return self._int64index._shallow_copy(values, **kwargs)
```
Comment (shoyer) [inline pandas/core/index.py]: You're dropping the `name` argument here. You want to use `kwargs['name'] = kwargs.get('name', self.name)` or something like that (I suppose better yet would be `kwargs.setdefault('name', self.name)`)

Comment (ARF1) [inline pandas/core/index.py]: Of course, thank you! Now I am finally getting somewhere with this. Only remaining issues are with `view` and `__getitem__()` using slices - which you already commented on on the other PR.

**Comment by shoyer** [comment]
@ARF1  Tracking down bugs like this can be unforgiving, especially when it takes you on a deep dive through pandas internals. Thanks for the hard work!

**Comment by ARF1** [comment]
Closing in favour of and rebasing changesets on #9977.

**Pull request closed without merging**